README.md
# telegram

[![NPM version](https://img.shields.io/npm/v/telegram.svg?style=flat)](https://npmjs.com/package/telegram)
[![NPM downloads](https://img.shields.io/npm/dm/telegram.svg?style=flat)](https://npmjs.com/package/telegram)
[![CircleCI](https://circleci.com/gh/zzswang/telegram/tree/master.svg?style=shield)](https://circleci.com/gh/zzswang/telegram/tree/master)
[![codecov](https://codecov.io/gh/zzswang/telegram/branch/master/graph/badge.svg)](https://codecov.io/gh/zzswang/telegram)
[![donate](https://img.shields.io/badge/$-donate-ff69b4.svg?maxAge=2592000&style=flat)](https://github.com/zzswang/donate)

## Install

```bash
yarn add telegram
```

## Usage

```js
const telegram = require("telegram");

telegram();
```

## API

### new Telegram()

Constructs a Telegram object. Returned object represents a parser which parses
nothing.

### decompress(buffer)

Parse a `Buffer` object `buffer` with this parser and return an object as result.

### [u]int{8, 16, 32}{le, be}(name[, options])

Parse bytes as an integer and store it in a variable named `name`. `name` should consist only of alphanumeric characters and start with an alphabet.

Number of bits can be chosen from 8, 16 and 32. Byte-ordering can be either `l` for little endian or `b` for big endian. With no prefix, it parses as a signed number, with `u` prefixed as an unsigned number.

```javascript
var parser = new Telegram()
  // Signed 32-bit integer (little endian)
  .int32le("a")
  // Unsigned 8-bit integer
  .uint8("b")
  // Signed 16-bit integer (big endian)
  .int16be("c");
```

### bit\[1-64\](name, options)

Parse bytes as a bit field and store it in variable `name`. There are 64 methods from `bit1` to `bit64` each corresponding to 1-bit-length to 64-bits-length bit field.

### {float, double}{le, be}(name[, options])

Parse bytes as an floating-point value and store it in a variable named `name`.

```javascript
var parser = new Telegram()
  // 32-bit floating value (big endian)
  .floatbe("a")
  // 64-bit floating value (little endian)
  .doublele("b");
```

### string(name, options)

Parse bytes as a string. `options` is an object which can have
the following keys:

- `encoding` - (Optional, defaults to `utf8`) Specify which encoding to use.
  `"utf8"`, `"ascii"`, `"hex"` and else are valid. See
  [`Buffer.toString`](http://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end)
  for more info.
- `length` - (Optional) Length of the string. Can be a number, string or a function. Use number for statically sized arrays.
- `zeroTerminated` - (Optional, defaults to `false`) If true, then this parser reads until it reaches zero.
- `greedy` - (Optional, defaults to `false`) If true, then this parser reads until it reaches the end of the buffer. Will consume zero-bytes.
- `stripNull` - (Optional, must be used with `length`) If true, then strip null characters from end of the string.

### array(name, options)

Parse bytes as an array. `options` is an object which can have the following keys:

- `type` - (Required) Type of the array element. Can be a string or an user defined Parser object. If it's a string, you have to choose from [u]int{8, 16, 32}{le, be} | string.
- `length` -
  Length of the array. Can be a number, string or a function. Use number for statically sized arrays.

```javascript
var parser = new Telegram()
  // Statically sized array
  .array("data", {
    type: "int32",
    length: 8,
  })

  // Dynamically sized array (references another variable)
  .uint8("dataLength")
  .array("data2", {
    type: "int32",
    length: "dataLength",
  })

  // Dynamically sized array (with some calculation)
  .array("data3", {
    type: "int32",
    length: function() {
      return this.dataLength - 1;
    }, // other fields are available through this
  });
```

### nest([name,] options)

Execute an inner parser and store its result to key `name`. If `name` is null
or omitted, the result of the inner parser is directly embedded into the
current object. `options` is an object which can have the following keys:

- `type` - (Required) A `Parser` object.

### skip(length[, options])

Skip parsing for `length` bytes. If `options` has key named `type` and the value of `type` is `"bit"`, skip parsing for `length` bits.

### endianess(endianess)

Define what endianess to use in this parser. `endianess` can be either
`"little"` or `"big"`. The default endianess of `Parser` is set to big-endian.

```javascript
var parser = new Telegram()
  .endianess("little")
  // You can specify endianess explicitly
  .uint16be("a")
  .uint32le("a")
  // Or you can omit endianess (in this case, little-endian is used)
  .uint16("b")
  .int32("c");
```

### Common options

These are common options that can be specified in all parsers.

- `formatter` - Function that transforms the parsed value into a more desired
  form.

  ```javascript
  var parser = new Telegram().array("ipv4", {
    type: uint8,
    length: "4",
    formatter: function(arr) {
      return arr.join(".");
    },
  });
  ```

- `assert` - Do assertion on the parsed result (useful for checking magic
  numbers and so on). If `assert` is a `string` or `number`, the actual parsed
  result will be compared with it with `===` (strict equality check), and an
  exception is thrown if they mismatch. On the other hand, if `assert` is a
  function, that function is executed with one argument (parsed result) and if
  it returns false, an exception is thrown.

  ```javascript
  // simple maginc number validation
  var ClassFile = Telegram.start()
    .endianess("big")
    .uint32("magic", { assert: 0xcafebabe });
  ```

## Examples

See `examples` for more complex examples.

## Contributing

1.  Fork it!
2.  Create your feature branch: `git checkout -b my-new-feature`
3.  Commit your changes: `git commit -am 'Add some feature'`
4.  Push to the branch: `git push origin my-new-feature`
5.  Submit a pull request :D

## Author

**telegram** © [zzswang](https://github.com/zzswang), Released under the [MIT](./LICENSE) License.

Authored and maintained by zzswang with help from contributors ([list](https://github.com/zzswang/telegram/contributors)).

> [github.com/zzswang](https://github.com/zzswang) · GitHub [@zzswang](https://github.com/zzswang)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.4.1](https://github.com/36node/telegram/compare/v0.4.0...v0.4.1) (2020-12-11)


### Bug Fixes

* update generate length logic for byte mode ([#18](https://github.com/36node/telegram/issues/18)) ([e8bc287](https://github.com/36node/telegram/commit/e8bc287))



## [0.4.0](https://github.com/36node/telegram/compare/v0.3.15...v0.4.0) (2020-12-10)


### Features

* add byte decompress mode ([#17](https://github.com/36node/telegram/issues/17)) ([82b1b05](https://github.com/36node/telegram/commit/82b1b05))



### [0.3.15](https://github.com/36node/telegram/compare/v0.3.14...v0.3.15) (2020-12-09)



### [0.3.14](https://github.com/36node/telegram/compare/v0.3.13...v0.3.14) (2020-12-07)



### [0.3.13](https://github.com/36node/telegram/compare/v0.3.12...v0.3.13) (2020-04-04)



### [0.3.12](https://github.com/36node/telegram/compare/v0.3.11...v0.3.12) (2020-03-16)



### [0.3.11](https://github.com/36node/telegram/compare/v0.3.10...v0.3.11) (2019-12-30)



### [0.3.10](https://github.com/36node/telegram/compare/v0.3.9...v0.3.10) (2019-07-16)


### Bug Fixes

* upgrade commitlint ([dd006e8](https://github.com/36node/telegram/commit/dd006e8))
* upgrade yarn lock ([1b59867](https://github.com/36node/telegram/commit/1b59867))
* use tpl-module template ([24a5385](https://github.com/36node/telegram/commit/24a5385))
* yarn upgrade ([5bcce41](https://github.com/36node/telegram/commit/5bcce41))



<a name="0.3.9"></a>
## [0.3.9](https://github.com/36node/telegram/compare/v0.3.8...v0.3.9) (2018-11-10)



<a name="0.3.8"></a>
## [0.3.8](https://github.com/36node/telegram/compare/v0.3.7...v0.3.8) (2018-11-10)


### Bug Fixes

* store before parse throw error ([70b4640](https://github.com/36node/telegram/commit/70b4640))



<a name="0.3.7"></a>
## [0.3.7](https://github.com/36node/telegram/compare/v0.3.6...v0.3.7) (2018-11-08)


### Bug Fixes

* log with err ([2bc6745](https://github.com/36node/telegram/commit/2bc6745))



<a name="0.3.6"></a>
## [0.3.6](https://github.com/36node/telegram/compare/v0.3.5...v0.3.6) (2018-11-07)


### Bug Fixes

* catch error in parser and set it into result ([c960f7a](https://github.com/36node/telegram/commit/c960f7a))



<a name="0.3.5"></a>
## [0.3.5](https://github.com/36node/telegram/compare/v0.3.4...v0.3.5) (2018-11-07)


### Bug Fixes

* catch error not quit ([0ecd06a](https://github.com/36node/telegram/commit/0ecd06a))



<a name="0.3.4"></a>
## [0.3.4](https://github.com/36node/telegram/compare/v0.3.3...v0.3.4) (2018-10-05)


### Bug Fixes

* es7 translate error with ...keys() ([3b19f1d](https://github.com/36node/telegram/commit/3b19f1d))



<a name="0.3.3"></a>
## [0.3.3](https://github.com/36node/telegram/compare/v0.3.2...v0.3.3) (2018-10-05)



<a name="0.3.2"></a>
## [0.3.2](https://github.com/36node/telegram/compare/v0.3.1...v0.3.2) (2018-10-05)



<a name="0.3.1"></a>
## [0.3.1](https://github.com/36node/telegram/compare/v0.3.0...v0.3.1) (2018-10-04)



<a name="0.3.0"></a>
# [0.3.0](https://github.com/36node/telegram/compare/v0.2.0...v0.3.0) (2018-09-06)


### Features

* array add readUntil ([e2f9428](https://github.com/36node/telegram/commit/e2f9428))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/36node/telegram/compare/v0.1.0...v0.2.0) (2018-09-04)


### Features

* skip bits & add testcases ([828e0ab](https://github.com/36node/telegram/commit/828e0ab))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/36node/telegram/compare/v0.0.2...v0.1.0) (2018-08-17)


### Bug Fixes

* markdownlint ([111cfe5](https://github.com/36node/telegram/commit/111cfe5))


### Features

* merge master ([5a262e7](https://github.com/36node/telegram/commit/5a262e7))
* parse string ([068e168](https://github.com/36node/telegram/commit/068e168))
* string can be array item ([9490f44](https://github.com/36node/telegram/commit/9490f44))
* string can be array item ([618de41](https://github.com/36node/telegram/commit/618de41))



<a name="0.0.2"></a>
## [0.0.2](https://github.com/36node/telegram/compare/v0.0.1...v0.0.2) (2018-08-06)



<a name="0.0.1"></a>
## [0.0.1](https://github.com/36node/telegram/compare/v0.1.1...v0.0.1) (2018-08-06)


src/lib.test.js
import { readBufferBits, readUIntBits, swapBuffer, writeBufferBits } from "./lib";

test("should readUIntBits", async () => {
  expect(readUIntBits(10, 4)).toEqual([1, 0, 1, 0]);

  expect(readUIntBits(14, 4)).toEqual([1, 1, 1, 0]);
  expect(readUIntBits(34, 6)).toEqual([1, 0, 0, 0, 1, 0]);

  const exp = [];
  exp.length = 53;
  exp.fill(1);
  expect(readUIntBits(Math.pow(2, 53) - 1, 53)).toEqual(exp);
});

test("should readerBufferBits", () => {
  const buf = Buffer.from("5d", "hex");

  const ret = readBufferBits(buf, 2, 3);

  expect(ret).toEqual([0, 1, 1]);
});

test("should writeBufferBits", () => {
  const buf = Buffer.from("5d", "hex");

  writeBufferBits(buf, [1, 0, 0], 2);

  expect(buf.toString("hex")).toEqual("65");
});

test("should swap buffer", () => {
  const buf1 = Buffer.from([1, 2, 3, 4]);
  swapBuffer(buf1, 1, 2);
  expect(buf1.equals(Buffer.from([1, 3, 2, 4])));

  const buf2 = Buffer.from([1, 2, 3, 4, 5, 6, 7]);
  swapBuffer(buf2, 2, 3);
  expect(buf2.equals(Buffer.from([1, 2, 5, 4, 3, 6, 7])));
});


src/parser.test.js
import fs from "fs";
import Telegram from "./parser";

const tcpHeader = new Telegram()
  .endianess("big")
  .uint16("srcPort")
  .uint16("dstPort")
  .uint32("seq")
  .uint32("ack")
  .bit4("dataOffset")
  .bit6("reserved")
  .nest("flags", {
    type: new Telegram()
      .bit1("urg")
      .bit1("ack")
      .bit1("psh")
      .bit1("rst")
      .bit1("syn")
      .bit1("fin"),
  })
  .uint16("windowSize")
  .uint16("checksum")
  .uint16("urgentPointer");

const tcpHeaderBuf = Buffer.from(
  "e8a203e108e177e13d20756b801829d3004100000101080a2ea486ba793310bc",
  "hex"
);

test("tcpHeader decompress", () => {
  expect(tcpHeader.decompress(tcpHeaderBuf)).toEqual({
    srcPort: 59554,
    dstPort: 993,
    seq: 148994017,
    ack: 1025537387,
    flags: { urg: 0, ack: 1, psh: 1, rst: 0, syn: 0, fin: 0 },
    dataOffset: 8,
    reserved: 0,
    windowSize: 10707,
    checksum: 65,
    urgentPointer: 0,
  });
});

const bmpFileHeader = new Telegram()
  .endianess("little")
  .string("type", {
    length: 2,
    assert: "BM",
  })
  .uint32("size")
  .uint16("reserved1")
  .uint16("reserved2")
  .uint32("offBits");

const bmpInfoHeader = new Telegram()
  .endianess("little")
  .uint32("size")
  .int32("width")
  .int32("height")
  .uint16("planes")
  .uint16("bitCount")
  .uint32("compression")
  .uint32("sizeImage")
  .int32("xPelsPerMeter")
  .int32("yPelsPerMeter")
  .uint32("clrUsed")
  .uint32("clrImportant");

const bmpFile = new Telegram()
  .nest("fileHeader", {
    type: bmpFileHeader,
  })
  .nest("infoHeader", {
    type: bmpInfoHeader,
  });

test("bmpFile1 decompress", () => {
  fs.readFile("examples/test.bmp", (err, data) => {
    if (err) throw err;
    expect(bmpFile.decompress(data)).toEqual({
      fileHeader: {
        type: "BM",
        size: 46182,
        reserved1: 0,
        reserved2: 0,
        offBits: 54,
      },
      infoHeader: {
        size: 40,
        width: 124,
        height: 124,
        planes: 1,
        bitCount: 24,
        compression: 0,
        sizeImage: 0,
        xPelsPerMeter: 0,
        yPelsPerMeter: 0,
        clrUsed: 0,
        clrImportant: 0,
      },
    });
  });
});

test("bmpFile2 decompress", () => {
  fs.readFile("examples/test2.bmp", (err, data) => {
    if (err) throw err;
    expect(bmpFile.decompress(data)).toEqual({
      fileHeader: {
        type: "BM",
        size: 151458,
        reserved1: 0,
        reserved2: 0,
        offBits: 138,
      },
      infoHeader: {
        size: 124,
        width: 259,
        height: 194,
        planes: 1,
        bitCount: 24,
        compression: 0,
        sizeImage: 151320,
        xPelsPerMeter: 3779,
        yPelsPerMeter: 3779,
        clrUsed: 0,
        clrImportant: 0,
      },
    });
  });
});

const oct2int = s => parseInt(s, 8);

const tarHeader = new Telegram()
  .string("name", { length: 100, stripNull: true })
  .string("mode", { length: 8, stripNull: true, formatter: oct2int })
  .string("uid", { length: 8, stripNull: true, formatter: oct2int })
  .string("gid", { length: 8, stripNull: true, formatter: oct2int })
  .string("size", { length: 12, stripNull: true, formatter: oct2int })
  .string("mtime", { length: 12, stripNull: true, formatter: oct2int })
  .string("chksum", { length: 8, stripNull: true, formatter: oct2int })
  .string("typeflag", { length: 1, stripNull: true, formatter: oct2int })
  .string("linkname", { length: 100, stripNull: true })
  .string("magic", { length: 6, stripNull: true })
  .string("version", { length: 2, stripNull: true, formatter: oct2int })
  .string("uname", { length: 32, stripNull: true })
  .string("gname", { length: 32, stripNull: true })
  .string("devmajor", { length: 8, stripNull: true, formatter: oct2int })
  .string("devminor", { length: 8, stripNull: true, formatter: oct2int })
  .string("prefix", { length: 155, stripNull: true })
  .skip(12);

const tarItem = new Telegram()
  .nest({
    type: tarHeader,
  })
  .skip(function(json) {
    return Math.ceil(json.size / 512) * 512;
  });

const tarArchive = new Telegram().array("files", {
  length: 8,
  type: tarItem,
});

test("tarFile decompress", () => {
  fs.readFile("examples/test.tar", (err, data) => {
    if (err) throw err;
    expect(tarArchive.decompress(data)).toEqual({
      files: [
        {
          name: "a/",
          mode: 493,
          uid: 501,
          gid: 20,
          size: 0,
          mtime: 1515642944,
          chksum: 5047,
          typeflag: 5,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/h",
          mode: 420,
          uid: 501,
          gid: 20,
          size: 44,
          mtime: 1515642944,
          chksum: 5152,
          typeflag: 0,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/c/",
          mode: 493,
          uid: 501,
          gid: 20,
          size: 0,
          mtime: 1515642936,
          chksum: 5199,
          typeflag: 5,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/b/",
          mode: 493,
          uid: 501,
          gid: 20,
          size: 0,
          mtime: 1515642920,
          chksum: 5196,
          typeflag: 5,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/b/e",
          mode: 420,
          uid: 501,
          gid: 20,
          size: 11,
          mtime: 1515642920,
          chksum: 5293,
          typeflag: 0,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/c/g",
          mode: 420,
          uid: 501,
          gid: 20,
          size: 33,
          mtime: 1515642936,
          chksum: 5299,
          typeflag: 0,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/c/d/",
          mode: 493,
          uid: 501,
          gid: 20,
          size: 0,
          mtime: 1515642929,
          chksum: 5346,
          typeflag: 5,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
        {
          name: "a/c/d/f",
          mode: 420,
          uid: 501,
          gid: 20,
          size: 22,
          mtime: 1515642929,
          chksum: 5448,
          typeflag: 0,
          linkname: "",
          magic: "ustar",
          version: 0,
          uname: "keichi",
          gname: "staff",
          devmajor: 0,
          devminor: 0,
          prefix: "",
        },
      ],
    });
  });
});


src/lib.js
/**
 * Returns an Array of length 8 containing the read bits.
 *
 * @example
 * readByteBits(42) => [0,0,1,0,1,0,1,0]
 *
 * @param {Number} byte one byte
 * @return {Array}
 */
export function readByteBits(byte) {
  if (byte > 255 || byte < 0 || ~~byte !== byte) throw new RangeError("invalid byte");
  const result = [0, 0, 0, 0, 0, 0, 0, 0];
  for (let i = 0; i < 8; i++) result[7 - i] = (byte >> i) & 1;
  return result;
}

/**
 * Return an array of length bitLength of read interger
 *
 * @example
 * readUnitBits(10, 4) => [1, 0, 1, 0]
 *
 * @param {number} val number to read
 * @param {number} bitLength max is 53
 */
export function readUIntBits(val, bitLength) {
  if (typeof val !== "number") throw new TypeError("val must be a number");

  if (val > Number.MAX_SAFE_INTEGER || val < 0) throw new RangeError("invalid number");

  const result = [];
  result.length = bitLength;
  result.fill(0);
  for (let i = 0; i < bitLength; i++) result[bitLength - 1 - i] = (val >> i) & 1;

  return result;
}

/**
 * Returns an Array containing bitLength bits starting at bitOffset.
 *
 * @example
 * readBufferBits(buffer, 2, 4) => [0,0,1,0]
 *
 * @param {Buffer} buffer the buffer to read
 * @param {Number} offset where to start (in bits)
 * @param {Number} length how many bits to read
 * @returns {Array}
 */
export function readBufferBits(buffer, offset = 0, length) {
  if (!length) length = buffer.length * 8 - offset;

  const start = Math.floor(offset / 8);
  const bytesToRead = Math.floor(length / 8) + 2;

  const arr = [];
  arr.length = bytesToRead * 8;

  for (let i = 0; i < bytesToRead; i++) {
    const toRead = buffer[start + i];
    if (toRead === undefined) continue;
    const bits = readByteBits(buffer[start + i]);
    arr[i * 8] = bits[0];
    arr[i * 8 + 1] = bits[1];
    arr[i * 8 + 2] = bits[2];
    arr[i * 8 + 3] = bits[3];
    arr[i * 8 + 4] = bits[4];
    arr[i * 8 + 5] = bits[5];
    arr[i * 8 + 6] = bits[6];
    arr[i * 8 + 7] = bits[7];
  }

  const subOffset = offset % 8;
  return arr.slice(subOffset, subOffset + length);
}

/**
 * Returns a UInt8 (0-255) which equals the given bits.
 *
 * @example
 * writeByteBits([0,0,1,0,1,0,1,0]) => 42
 *
 * @param {Array} byte 8 bits
 * @return {Number} 8-bit unsigned integer
 */
export function writeByteBits(byte) {
  if (!Array.isArray(byte) || byte.length !== 8) throw new RangeError("invalid array length");

  let data = 0;

  for (let i = 0; i < 8; i++) if (byte[7 - i]) data |= 1 << i;

  return data;
}

/**
 * Modifies the buffer's bits to equal newBits starting at bitOffset.
 *
 * @example
 * writeBufferBits(buffer, [0,0,1,0], 0) => buffer was modified
 *
 * @param {Buffer} buffer the buffer to modify
 * @param {Array} bits the bits to insert
 * @param {Number} offset where to start (in bits)
 * @returns {undefined}
 */
export function writeBufferBits(buffer, bits, offset) {
  const start = Math.floor(offset / 8);
  const end = Math.ceil((offset + bits.length) / 8);

  const subBuffer = buffer.slice(start, end);
  const byteData = readBufferBits(subBuffer);

  let subOffset = offset % 8;

  for (var i = 0; i < bits.length; i++) byteData[subOffset++] = bits[i];

  const length = end - start;
  for (let i = 0; i < length; i++) subBuffer[i] = writeByteBits(byteData.slice(i * 8, (i + 1) * 8));
}

/**
 * swap buffer endian starting at offset
 *
 * @example
 * swapBuffer([0x1, 0x2, 0x3, 0x4], 1, 2)  => [0x1, 0x3, 0x2, 0x4]
 *
 * @param {Buffer} buffer the buffer to swap
 * @param {number} offset the byte to start
 * @param {number} length how many bytes to swap
 */
export function swapBuffer(buffer, offset, length) {
  for (let i = 0; i < Math.floor(length / 2); i++) {
    const left = i + offset;
    const right = offset + length - 1 - i;

    const tmp = buffer[left];
    buffer[left] = buffer[right];
    buffer[right] = tmp;
  }
}


examples/bmp.js
import fs from "fs";
import Telegram from "../src";

// C structure BITMAPFILEHEADER
// typedef struct tagBITMAPFILEHEADER {
//   WORD  bfType;
//   DWORD bfSize;
//   WORD  bfReserved1;
//   WORD  bfReserved2;
//   DWORD bfOffBits;
// } BITMAPFILEHEADER, *PBITMAPFILEHEADER;
const bmpFileHeader = new Telegram()
  .endianess("little")
  .string("type", {
    length: 2,
    assert: "BM",
  })
  .uint32("size")
  .uint16("reserved1")
  .uint16("reserved2")
  .uint32("offBits");

// C structure BITMAPINFOHEADER definition
// typedef struct tagBITMAPINFOHEADER {
//     DWORD  biSize;
//     LONG   biWidth;
//     LONG   biHeight;
//     WORD   biPlanes;
//     WORD   biBitCount;
//     DWORD  biCompression;
//     DWORD  biSizeImage;
//     LONG   biXPelsPerMeter;
//     LONG   biYPelsPerMeter;
//     DWORD  biClrUsed;
//     DWORD  biClrImportant;
// } BITMAPINFOHEADER;
const bmpInfoHeader = new Telegram()
  .endianess("little")
  .uint32("size")
  .int32("width")
  .int32("height")
  .uint16("planes")
  .uint16("bitCount")
  .uint32("compression")
  .uint32("sizeImage")
  .int32("xPelsPerMeter")
  .int32("yPelsPerMeter")
  .uint32("clrUsed")
  .uint32("clrImportant");

const bmpFile = new Telegram()
  .nest("fileHeader", {
    type: bmpFileHeader,
  })
  .nest("infoHeader", {
    type: bmpInfoHeader,
  });

fs.readFile("examples/test.bmp", (err, data) => {
  if (err) throw err;
  console.log(bmpFile.decompress(data));
});


examples/jpg.js
import fs from "fs";
import Telegram from "../src";

const SOI = new Telegram();

const EOI = new Telegram();

const APP0 = new Telegram()
  .endianess("big")
  .uint16("length")
  .string("id", {
    encoding: "ascii",
    zeroTerminated: true,
    validate: "JFIF",
  })
  .uint16("version")
  .uint8("unit")
  .uint16("xDensity")
  .uint16("yDensity")
  .uint8("thumbWidth")
  .uint8("thumbHeight")
  .array("thumbData", {
    type: "uint8",
    length: function(json) {
      return json.Xt * json.Yt * 3;
    },
  });

const COM = new Telegram()
  .endianess("big")
  .uint16("length")
  .string("comment", {
    encoding: "ascii",
    length: function(json) {
      return json.length - 2;
    },
  });

const SOS = new Telegram()
  .endianess("big")
  .uint16("length")
  .uint8("componentCount")
  .array("components", {
    type: new Telegram().uint8("id").uint8("dht"),
    length: "componentCount",
  })
  .uint8("spectrumStart")
  .uint8("spectrumEnd")
  .uint8("spectrumSelect");

const DQT = new Telegram()
  .endianess("big")
  .uint16("length")
  .array("tables", {
    type: new Telegram().uint8("precisionAndTableId").array("table", {
      type: "uint8",
      length: 64,
    }),
    length: function(json) {
      return (json.length - 2) / 65;
    },
  });

const SOF0 = new Telegram()
  .endianess("big")
  .uint16("length")
  .uint8("precision")
  .uint16("width")
  .uint16("height")
  .uint8("componentCount")
  .array("components", {
    type: new Telegram()
      .uint8("id")
      .uint8("samplingFactor")
      .uint8("quantizationTableId"),
    length: "componentCount",
  });

const Ignore = new Telegram()
  .endianess("big")
  .uint16("length")
  .skip(function(json) {
    return json.length - 2;
  });

const switchMarker = json => {
  switch (json.marker) {
    case 0xffd8:
      return SOI;
    case 0xffd9:
      return EOI;
    case 0xffe0:
      return APP0;
    case 0xffda:
      return SOS;
    case 0xffdb:
      return DQT;
    case 0xffc0:
      return SOF0;
    default:
      return Ignore;
  }
};

const Segment = new Telegram()
  .endianess("big")
  .uint16("marker")
  .nest("segment", { type: switchMarker });

const JPEG = new Telegram().array("segments", {
  type: Segment,
  length: 9,
});

fs.readFile("examples/test.jpg", function(err, data) {
  console.dir(JPEG.decompress(data), { depth: null, colors: true });
  // console.log(require("util").inspect(JPEG.decompress(data), { depth: null }));
});


examples/tar.js
import fs from "fs";
import Telegram from "../src";

const oct2int = s => parseInt(s, 8);

const tarHeader = new Telegram()
  .string("name", { length: 100, stripNull: true })
  .string("mode", { length: 8, stripNull: true, formatter: oct2int })
  .string("uid", { length: 8, stripNull: true, formatter: oct2int })
  .string("gid", { length: 8, stripNull: true, formatter: oct2int })
  .string("size", { length: 12, stripNull: true, formatter: oct2int })
  .string("mtime", { length: 12, stripNull: true, formatter: oct2int })
  .string("chksum", { length: 8, stripNull: true, formatter: oct2int })
  .string("typeflag", { length: 1, stripNull: true, formatter: oct2int })
  .string("linkname", { length: 100, stripNull: true })
  .string("magic", { length: 6, stripNull: true })
  .string("version", { length: 2, stripNull: true, formatter: oct2int })
  .string("uname", { length: 32, stripNull: true })
  .string("gname", { length: 32, stripNull: true })
  .string("devmajor", { length: 8, stripNull: true, formatter: oct2int })
  .string("devminor", { length: 8, stripNull: true, formatter: oct2int })
  .string("prefix", { length: 155, stripNull: true })
  .skip(12);

const tarItem = new Telegram()
  .nest({
    type: tarHeader,
  })
  .skip(function(json) {
    return Math.ceil(json.size / 512) * 512;
  });

const tarArchive = new Telegram().array("files", {
  length: 8,
  type: tarItem,
});

fs.readFile("examples/test.tar", function(err, data) {
  console.dir(tarArchive.decompress(data), { depth: null, colors: true });
});
